cpu-o3: refactored load replay&cancel, clean code

tastynoob opened the pull request:
[WIP]
finished:
Move all replays to loadpipe s2 (actually in s3, decided by function calling order)
Move all cancels to loadpipe s3
[WARN]
still debugging, just for ci test

tastynoob committed:
cpu-o3: refactored load replay&cancel, clean code

tastynoob committed:
cpu-o3: fix strictly ordered

Change-Id: I809dc306702de911a2561043cd59e25cff34fdb4

tastynoob committed:
cpu-o3: kmhv3 enable nuke replay, miss replay

Change-Id: I9e116ed6149837dc410b37e63ae77d8f530ab79a

fix

tastynoob committed:
fix kmhv3 request delete untimely

fix

happy-lx commented on src/cpu/o3/lsq_unit.cc:
```diff
@@ -1190,164 +1139,127 @@ LSQUnit::loadPipeS1(const DynInstPtr &inst, std::bitset<LdStFlagNum> &flag)
+            inst->setExecuted();
+        }
+        inst->setSkipFollowingPipe();
+        if (inst->needReplay()) {
```
maybe `!inst->needReplay()`

happy-lx commented on src/cpu/o3/lsq.cc:
```diff
@@ -1065,20 +1067,23 @@ LSQ::pushRequest(const DynInstPtr& inst, bool isLoad, uint8_t *data,
+        return std::make_shared<RiscvISA::AddressFault>(request->mainReq()->getVaddr(),
+            request->mainReq()->getgPaddr(),
+            RiscvISA::ExceptionCode::STORE_ADDR_MISALIGNED);
+    }
```
load & store & atomic
maybe can write like
```c
if (isMisaligned(inst, request)) {
    return inst->getFault();
}
```

happy-lx committed:
cpu-o3: fix strictlyOrdered load stuck

now, strictlyOrdered load does not writeback to ROB using `readyToFinish`,
so set it as `CanCommit` the first time in case rob can not commit this

Change-Id: If9ff72bb9e22ec1796682ecd879d0f2ecfab4aeb

Llemonade committed:
cpu-o3: fix misalign check bug

Change-Id: Ie3a65514abc1573a7c1abb6ece5ec349ee3ce079

Ergou-ren committed:
cpu-o3: Fixed logical errors caused by refactoring

Change-Id: I325e4852ff4b257ed723bb66ee6aad92cc46d611

Ergou-ren committed:
cpu-o3: Add handling for cache misses with a latency of 1.

When cache prefetching is enabled, a cache miss with a latency of 1 can occur, which is a special type of cache hit.

Change-Id: I0daf5933ce61274eccf77c10ef148ac7f2aec31b

Ergou-ren committed:
cpu-o3: Add a flag for cache hits in the MSHR.

cacheSatisfiedByMSHR is used to record cases where the request can be satisfied by the MSHR.

Change-Id: I237d39006fd4108907b71ef8d18f69f02185dcfb

Ergou-ren committed:
cpu-o3: Modify the replay stage

Modify fast replay from s3 to s2.

Change-Id: I755020b0b1533f0fd61b4a19a59d00929a6a8064

Ergou-ren committed:
cpu-o3: check cache hit or miss when receiving the response.

Change-Id: I79d7f68ad3b01e4fdd4056ec695d374e1358d034

Ergou-ren commented on src/cpu/o3/iew.cc:
```diff
@@ -2098,7 +2098,7 @@ IEW::checkLoadStoreInst(DynInstPtr inst)
     if (inFlight) {
         assert(inst->pendingCacheReq);
-        depth = inst->pendingCacheReq->mainReq()->depth;
+        depth = 0;// inst->pendingCacheReq->mainReq()->depth;
```
@tastynoob The comment can delete ？

Ergou-ren replied:
@tastynoob The comment can delete ？

Ergou-ren commented on src/cpu/o3/lsq_unit.cc:
```diff
@@ -857,10 +864,10 @@ LSQUnit::checkSnoop(PacketPtr pkt)
     DynInstPtr ld_inst = iter->instruction();
     assert(ld_inst);
-    LSQRequest *request = iter->request();
+    LSQRequest *request = ld_inst->savedRequest; //iter->request();
```
@tastynoob look at can delete

Ergou-ren committed:
cpu-o3: Format the code.

Change-Id: I2179671ad03fae86c2ec212c5edadbf4c96d88c6
